README.md
A Flutter package that simplify the internationalizing process using JSON and YAML files. Extracts messages to generate Dart files with friendly way to access messages you need.

`String get translator => "tradutor";`

[![Pub](https://img.shields.io/pub/v/tradutor?color=blueviolet)](https://pub.dev/packages/tradutor)

## Features

* Simple messages.
* Messages with multiple parameters.
* List of simple messages.
* List of messages with multiple parameters.
* Supported message types:
  * Plural.
  * Date.
  * Numeric.
  * Map.
* Optional parameters support.
* JSON and YAML files.
* Nested messages support.
* Supports hot reload.

## Installation 

In `pubspec.yaml` add the following dependencies:

```yaml
dependencies: 
    flutter_localizations: 
        sdk: flutter 

dev_dependencies:
    tradutor: ^0.10.0
```

## Usage
Open a terminal inside your project directory and run the following command:

```
flutter pub run tradutor:generate
```

All supported arguments:

* `--source "DIRECTORY PATH"` or `-s "DIRECTORY PATH"`: A source folder contains all JSON/YAML files (defaults to "./i18n");
* `--output "FILE PATH"` or `-o "FILE PATH"`: An output file contains all strings (defaults to "./lib/i18n.dart");
* `--fallback "LANGUAGE"` or `-f  "LANGUAGE"`:  Provides a default language, used when the translation for the current running system is not provided (defaults to "en_US");
* `--watch`: Watches the JSON/YAML files for edits and does rebuilds as necessary.
* `--class-name "NAME"` or `-c "NAME"`: Allows change the generated Dart class name (defaults to "I18n").

Full example: `flutter packages pub run tradutor:generate -s "/i18n" -o "/lib/i18n.dart" -f "en_US" -c "I18n" --watch`

## JSON Files
Crete JSON files naming them with language code (lowercase) and country code (uppercase). Ex.: `pt_BR.json`, `en_US.json`, etc.

`en_US.json`
```json
{
    "simpleMessage": "This is a simple Message",
    "messageWithParameters": "Hi {name}, Welcome you!",
    "brazilFlagColors": ["Green", "Yellow", "Blue", "White"],
    "simpleWhiteCakeIngredients": [
        "{whiteSugar} cup white sugar",
        "{butter} cup butter",
        "{eggs} eggs",
        "{vanilla} teaspoons vanilla extract",
        "{flour} cups all-purpose flour",
        "{bakingPowder} teaspoons baking powder",
        "{milk} cup milk"
    ],
    "homePage": {
        "title": "Home Page"
    },
    "counter.one": "Button clicked 1 time",
    "counter.other": "Button cliked {quantity} times"
}
```

## YAML Files
Crete YAML files naming them with language code (lowercase) and country code (uppercase). Ex.: `pt_BR.yaml`, `en_US.yaml`, etc.

```yaml
simpleMessage: これは簡単なメッセージです
messageWithParameters: "{name}様、ようこそ！"
brazilFlagColors:
- 緑
- 黄色
- 青い
- 白い
simpleWhiteCakeIngredients:
- "白砂糖{whiteSugar}カップ"
- "バター{butter}カップ"
- "卵{eggs}個"
- "バニラエッセンス小さじ{vanilla}"
- "薄力粉{flour}カップ"
- "小さじ{bakingPowder}杯のベーキングパウダー"
- "牛乳{milk}カップ"
homePage:
  title: ホームページ
counter.other: ボタンが{quantity}回クリックされた
```

## Nested messages

The messages can be nested. The final message name will be transformed to camelCase format (starting with a lower case character and capital for each later key).

```json
{
  "this" : {
    "is" : {
      "a": {
        "nested": "Message"
      }
    }
  }
}
```

Generated Dart getter:

```dart
String get thisIsANested => 'Message';
```

## Optional parameters

Use `{?name=value}` for set the parameter as optional. The `=value` is optional and will use an empty string when omitted.

## Supported Message Type

The all parameters in a message must be surrounded by curly braces. The parameters are sorted alphabetically.

### Simple message

```json
{ 
    "simpleMessage": "This is a simple Message"
}
```

Generated Dart getter:

```dart
String get simpleMessage => 'This is a simple Message';
```

### Message with parameters

```json
{
    "messageWithParameters": "Hi {name}, Welcome you!"
}
```

Generated Dart method:

```dart
String messageWithParameters(dynamic name) => 'Hi ${name}, Welcome you!';
```

### List of simple messages

Define an array of strings.

```json
{
    "brazilFlagColors": ["Green", "Yellow", "Blue", "White"]
}
```

Generated Dart getter:

```dart
List<String> get brazilFlagColors => ['Green', 'Yellow', 'Blue', 'White'];
```

### List of messages with multiple parameters

```json
{
    "simpleWhiteCakeIngredients": [
        "{whiteSugar} cup white sugar",
        "{butter} cup butter",
        "{eggs} eggs",
        "{vanilla} teaspoons vanilla extract",
        "{flour} cups all-purpose flour",
        "{bakingPowder} teaspoons baking powder",
        "{milk} cup milk"
    ],
}
```

Generated Dart method:

```dart
List<String> simpleWhiteCakeIngredients(
          dynamic bakingPowder, 
          dynamic butter, 
          dynamic eggs, 
          dynamic flour, 
          dynamic milk, 
          dynamic vanilla, 
          dynamic whiteSugar) =>
      [
        '${whiteSugar} cup white sugar',
        '${butter} cup butter',
        '${eggs} eggs',
        '${vanilla} teaspoons vanilla extract',
        '${flour} cups all-purpose flour',
        '${bakingPowder} teaspoons baking powder',
        '${milk} cup milk'
      ];
```

### Plural messages

The last message name must be `zero`, `one`, `two`, `few`, `many` or `other`. See [Language Plural Rules](https://unicode-org.github.io/cldr-staging/charts/37/supplemental/language_plural_rules.html).

```json
{
  "counter.one": "Button clicked 1 time",
  "counter.other": "Button cliked {quantity} times"
}
```

Or

```json
{
  "counter": {
    "one": "Button clicked 1 time",
    "other": "Button cliked {quantity} times"
  }
}
```

> The 'other' plural form must be provided.

> `quantity` is the parameter used to format a message depending on its value.

Generated Dart method:

```dart
String counter(int quantity) => Intl.plural(
        quantity,
        locale: 'en_US',
        one: 'Button clicked 1 time',
        other: 'Button cliked ${quantity} times',
      );
```

### Map messages

The message name starts with `%`.

```json
{
  "%eyeColor": {
    "FAMALE": "She has {color} eyes",
    "MALE": "He has {color} eyes"
  }
}
```

Generated Dart method:

```dart
String eyeColor(String key, dynamic color) {
    switch (key) {
      case 'FAMALE':
        return 'She has ${color} eyes';
      case 'MALE':
        return 'He has ${color} eyes';
      default:
        return null;
    }
  }
```

> `key` is the parameter used to format a message depending on its value.

### Date messages

The message name starts with `#`.

```json
{
    "#fullDate": "MM dd, yyyy h:mm:ss a"
}
```

Generated Dart method:

```dart
static final _fullDateFormatter = DateFormat('MM dd, yyyy h:mm:ss a', 'en_US');
String fullDate(DateTime date) => _fullDateFormatter.format(date);
```

See the DateFormat documentation [here](https://pub.dev/documentation/intl/latest/intl/DateFormat-class.html).

### Number messages

The message name starts with `$`.

```json
{
    "$currencyMessage": "\\$#,##0.00"
}
```

Generated Dart method:

```dart
static final _currencyMessageFormatter = NumberFormat('\$#,##0.00', 'en_US');
String currencyMessage(num number) => _currencyMessageFormatter.format(number);
```

See the NumberFormat documentation [here](https://pub.dev/documentation/intl/latest/intl/NumberFormat-class.html).

### Usage of Generated Dart file (i18n.dart)

```dart
import 'package:flutter/material.dart';
import 'package:flutter_localizations/flutter_localizations.dart';

import 'i18n.dart';

void main() => runApp(App());

class App extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final i18n = I18n.delegate;

    return MaterialApp(
      title: 'Tradutor',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const HomePage(),
      localizationsDelegates: [
        i18n,
        GlobalMaterialLocalizations.delegate,
        GlobalWidgetsLocalizations.delegate,
      ],
      supportedLocales: i18n.supportedLocales,
      localeResolutionCallback: i18n.resolution(
        fallback: const Locale('en', 'US'),
      ),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({
    Key key,
  }) : super(key: key);

  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  var _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    final i18n = I18n.of(context);

    return Scaffold(
      appBar: AppBar(
        title: Text(i18n.homePage),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(i18n.counter(_counter)),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        child: const Icon(Icons.add),
      ),
    );
  }
}
```

## Escaping Curly Braces and Dollar Sign.

| Json | Dart | Message (a = 'Hi') | Note |
| :-: | :-: | :-: | :-: |
| "{a}" | "${a}" | "Hi" | Just a parameter |
| "\\\\{a}" | "\\{a}"  | "{a}" | Text surrounded by curly braces |
| "\\\\\\\\{a}"  | "\\\\{a}" | "\\{a}" | Backslash + text surrounded by curly braces |
| "\\\\{!a}" | "\\${a}" | "\\Hi" | Backslash + parameter |
| "\\\\{!?a}" | "\\${a}" | "\\Hi" | Backslash + optional parameter |
| "\\\\$"  | "\\$" | "$" | Just a dollar sign |

## Language Options

### @parent

Use `@parent` to indicate that language is a parent language.

`pt_PT.json`
```json
{
    "@parent": true,
}
```

`pt_PT` is parent language of `pt_BR`, `pt_AO`, `pt_TL`, etc. If some child language does not found, `pt_PT` will be used instead.

### @textDirection

Use `@textDirection` to indicate the direction in which text flows. The valid values are `ltr` (default) and `rtl`.

`he_IL.json`
```json
{
    "@textDirection": "rtl",
}
```


CHANGELOG.md
## [0.10.0]
 - Null safety support.

## [0.9.2]
 - Bug fix.

## [0.9.1]
 - Bug fix.

## [0.9.0]
 - Add Optional parameter support.

## [0.8.0]
 - Add Number Format support.
 - Add Map support.
 - Various bug fixes.

## [0.7.0]
 - Code refactoring.
 - Otimizations and improved performance.
 - Fix escape curly braces.
 - [Breaking]: Remove `@extendsOf` property in favor of the `@parent` property.
 - [Breaking]: Remove web support (temporarily).

## [0.6.3]
 - Ignore messages with null value.

## [0.6.2]
 - Bug fixes.

## [0.6.1]
 - Bug fixes.

## [0.6.0]
 - Add Date message type.
 - Bug fixes.

## [0.5.1]
 - Bug fixed.

## [0.5.0]
 - Add support to Flutter Web.
 - Add widget tests.

## [0.4.0]

 - Add support to YAML files.

## [0.3.0]

 - Add support to change the generated Dart class name.

## [0.2.0]

 - Add support to nested messages.
 - Bug fixes.

## [0.1.0]

 - Initial release.


example/lib/i18n.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';
import 'package:intl/intl.dart' hide TextDirection;

// ignore_for_file: camel_case_types
// ignore_for_file: non_constant_identifier_names
// ignore_for_file: unnecessary_brace_in_string_interps
// ignore_for_file: unused_import
// ignore_for_file: annotate_overrides
// ignore_for_file: avoid_annotating_with_dynamic
// ignore_for_file: implicit_dynamic_parameter

// See more about language plural rules: https://unicode-org.github.io/cldr-staging/charts/37/supplemental/language_plural_rules.html

const I18n enUS = _I18n_en_US();
const I18n ptBR = _I18n_pt_BR();
const I18n ptPT = _I18n_pt_PT();

class _I18n_en_US extends I18n {
  const _I18n_en_US();
}

class I18n implements WidgetsLocalizations {
  const I18n();

  static Locale _locale;

  static bool _shouldReload = false;

  static const GeneratedLocalizationsDelegate delegate =
      GeneratedLocalizationsDelegate();

  static final _dateMessageFormatter =
      DateFormat('MM dd, yyyy h:mm:ss a', 'en_US');

  static final _moneyMessageFormatter = NumberFormat('\$#,##0.00', 'en_US');

  static final _numberMessageFormatter = NumberFormat('###.0#', 'en_US');

  static final _percentMessageFormatter = NumberFormat('###.0%', 'en_US');

  static Locale get locale$ => _locale;
  static set locale$(Locale locale) {
    _shouldReload = true;
    _locale = locale;
  }

  TextDirection get textDirection => TextDirection.ltr;
  static I18n of(BuildContext context) =>
      Localizations.of<I18n>(context, WidgetsLocalizations);
  String constellations(String key, dynamic star) {
    switch (key) {
      case 'AND':
        return 'Andromeda (${key}) - Brightest star: ${star}';
      case 'ANT':
        return 'Antlia (${key}) - Brightest star: ${star}';
      case 'APS':
        return 'Apus (${key}) - Brightest star: ${star}';
      default:
        return null;
    }
  }

  String dateMessage(DateTime date) => _dateMessageFormatter.format(date);
  String escapeMessage(dynamic b, dynamic d, dynamic e) =>
      '{a} ${b} \\{c} \\${d} {${e}} {{f}} \$';
  List<String> get listMessage => ['Green', 'Yellow', 'Blue', 'White'];
  List<String> listMessageWithParameters(
          dynamic bakingPowder,
          dynamic butter,
          dynamic eggs,
          dynamic flour,
          dynamic milk,
          dynamic vanilla,
          dynamic whiteSugar) =>
      [
        '${whiteSugar} cup white sugar',
        '${butter} cup butter',
        '${eggs} eggs',
        '${vanilla} teaspoons vanilla extract',
        '${flour} cups all-purpose flour',
        '${bakingPowder} teaspoons baking powder',
        '${milk} cup milk'
      ];
  String get messageKeyCamelCase => 'This is a message';
  String moneyMessage(num number) => _moneyMessageFormatter.format(number);
  String get notTranslatable => 'This message is not translatable';
  String numberMessage(num number) => _numberMessageFormatter.format(number);
  String optionalParameter({dynamic name}) => 'You are ${name ?? "Anonymous"}';
  String percentMessage(num number) => _percentMessageFormatter.format(number);
  String pluralMesssage(int quantity) => Intl.plural(quantity,
      locale: 'en_US',
      one: 'Button tapped ${quantity} time',
      other: 'Button tapped ${quantity} times');
  String pluralMesssageGrouped(int quantity) => Intl.plural(quantity,
      locale: 'en_US',
      one: 'Button tapped ${quantity} time',
      other: 'Button tapped ${quantity} times');
  String pluralMesssageWithParameters(
          int quantity, dynamic action, dynamic item) =>
      Intl.plural(quantity,
          locale: 'en_US',
          one: '${item} ${action} ${quantity} time',
          other: '${item} ${action} ${quantity} times');
  String get simpleMessage => 'Home Page';
  String simpleMessageWithParameters(dynamic name) => 'Hello ${name}';
}

class _I18n_pt_BR extends I18n {
  const _I18n_pt_BR();

  static final _dateMessageFormatter =
      DateFormat('dd/MM/yyyy HH:mm:ss', 'pt_BR');

  static final _moneyMessageFormatter = NumberFormat('R\$ #,##0.00', 'pt_BR');

  static final _numberMessageFormatter = NumberFormat('###.0#', 'pt_BR');

  static final _percentMessageFormatter = NumberFormat('###.0%', 'pt_BR');

  TextDirection get textDirection => TextDirection.ltr;
  String constellations(String key, dynamic star) {
    switch (key) {
      case 'AND':
        return 'Andrômeda (${key}) - Estrela mais brilhante: ${star}';
      case 'ANT':
        return 'Máquina Pneumática (${key}) - Estrela mais brilhante: ${star}';
      case 'APS':
        return 'Ave-do-paraíso (${key}) - Estrela mais brilhante: ${star}';
      default:
        return null;
    }
  }

  String dateMessage(DateTime date) => _dateMessageFormatter.format(date);
  List<String> get listMessage => ['Verde', 'Amarelo', 'Azul', 'Branco'];
  List<String> listMessageWithParameters(
          dynamic bakingPowder,
          dynamic butter,
          dynamic eggs,
          dynamic flour,
          dynamic milk,
          dynamic vanilla,
          dynamic whiteSugar) =>
      [
        '${whiteSugar} copos de açúcar cristal',
        '${butter} copo de manteiga',
        '${eggs} ovos',
        '${vanilla} colher (chá) de extrato de baunilha',
        '${flour} copos de farinha de trigo',
        '${bakingPowder} colher (chá) de fermento em pó',
        '${milk} copo de leite'
      ];
  String get messageKeyCamelCase => 'Isto é uma mensagem';
  String moneyMessage(num number) => _moneyMessageFormatter.format(number);
  String numberMessage(num number) => _numberMessageFormatter.format(number);
  String optionalParameter({dynamic name}) => 'Você é ${name ?? "Anônimo"}';
  String percentMessage(num number) => _percentMessageFormatter.format(number);
  String pluralMesssage(int quantity) => Intl.plural(quantity,
      locale: 'pt_BR',
      one: 'Botão foi clicado ${quantity} vez',
      other: 'Botão foi clicado ${quantity} vezes');
  String pluralMesssageWithParameters(
          int quantity, dynamic action, dynamic item) =>
      Intl.plural(quantity,
          locale: 'pt_BR',
          one: '${item} foi ${action} ${quantity} vez',
          other: '${item} foi ${action} ${quantity} vezes');
  String get simpleMessage => 'Página Inicial';
  String simpleMessageWithParameters(dynamic name) => 'Olá ${name}';
}

class _I18n_pt_PT extends I18n {
  const _I18n_pt_PT();

  static final _dateMessageFormatter =
      DateFormat('dd/MM/yyyy HH:mm:ss', 'pt_PT');

  static final _moneyMessageFormatter = NumberFormat('#,##0.00 €', 'pt_PT');

  static final _numberMessageFormatter = NumberFormat('###.0#', 'pt_PT');

  static final _percentMessageFormatter = NumberFormat('###.0%', 'pt_PT');

  TextDirection get textDirection => TextDirection.ltr;
  String constellations(String key, dynamic star) {
    switch (key) {
      case 'AND':
        return 'Andrômeda (${key}) - Estrela mais brilhante: ${star}';
      case 'ANT':
        return 'Máquina Pneumática (${key}) - Estrela mais brilhante: ${star}';
      case 'APS':
        return 'Ave-do-paraíso (${key}) - Estrela mais brilhante: ${star}';
      default:
        return null;
    }
  }

  String dateMessage(DateTime date) => _dateMessageFormatter.format(date);
  List<String> get listMessage => ['Verde', 'Amarelo', 'Azul', 'Branco'];
  List<String> listMessageWithParameters(
          dynamic bakingPowder,
          dynamic butter,
          dynamic eggs,
          dynamic flour,
          dynamic milk,
          dynamic vanilla,
          dynamic whiteSugar) =>
      [
        '${whiteSugar} copos de açúcar cristal',
        '${butter} copo de manteiga',
        '${eggs} ovos',
        '${vanilla} colher (chá) de extrato de baunilha',
        '${flour} copos de farinha de trigo',
        '${bakingPowder} colher (chá) de fermento em pó',
        '${milk} copo de leite'
      ];
  String get messageKeyCamelCase => 'Isto é uma mensagem';
  String moneyMessage(num number) => _moneyMessageFormatter.format(number);
  String numberMessage(num number) => _numberMessageFormatter.format(number);
  String optionalParameter({dynamic name}) => 'Você é ${name ?? "Anônimo"}';
  String percentMessage(num number) => _percentMessageFormatter.format(number);
  String pluralMesssage(int quantity) => Intl.plural(quantity,
      locale: 'pt_PT',
      one: 'Botão foi pressionado ${quantity} vez',
      other: 'Botão foi pressionado ${quantity} vezes');
  String pluralMesssageWithParameters(
          int quantity, dynamic action, dynamic item) =>
      Intl.plural(quantity,
          locale: 'pt_PT',
          one: '${item} foi ${action} ${quantity} vez',
          other: '${item} foi ${action} ${quantity} vezes');
  String get simpleMessage => 'Página Inicial';
  String simpleMessageWithParameters(dynamic name) => 'Olá ${name}';
}

class GeneratedLocalizationsDelegate
    extends LocalizationsDelegate<WidgetsLocalizations> {
  const GeneratedLocalizationsDelegate();

  List<Locale> get supportedLocales =>
      const [Locale('en', 'US'), Locale('pt', 'BR'), Locale('pt', 'PT')];
  LocaleResolutionCallback resolution({Locale fallback}) {
    return (locale, supported) {
      return isSupported(locale) ? locale : (fallback ?? supported.first);
    };
  }

  Future<WidgetsLocalizations> load(Locale locale) {
    I18n._locale ??= locale;
    I18n._shouldReload = false;
    locale = I18n._locale;
    final lang = locale?.toString() ?? '';
    final languageCode = locale?.languageCode ?? '';
    if ('en_US' == lang) {
      return SynchronousFuture<WidgetsLocalizations>(const _I18n_en_US());
    }
    if ('pt_BR' == lang) {
      return SynchronousFuture<WidgetsLocalizations>(const _I18n_pt_BR());
    }
    if ('pt_PT' == lang) {
      return SynchronousFuture<WidgetsLocalizations>(const _I18n_pt_PT());
    }
    if ('en' == languageCode) {
      return SynchronousFuture<WidgetsLocalizations>(const _I18n_en_US());
    }
    if ('pt' == languageCode) {
      return SynchronousFuture<WidgetsLocalizations>(const _I18n_pt_PT());
    }
    return SynchronousFuture<WidgetsLocalizations>(const I18n());
  }

  bool isSupported(Locale locale) {
    for (var i = 0; i < supportedLocales.length && locale != null; i++) {
      final l = supportedLocales[i];
      if (l.languageCode == locale.languageCode) {
        return true;
      }
    }
    return false;
  }

  bool shouldReload(GeneratedLocalizationsDelegate old) => I18n._shouldReload;
}
